### PR TITLE
Disable widgets

### PR DESCRIPTION
**Pull request by xarvic**
An implementation of a disabled state for widgets (issue: #746 ).
It adds the following:

- `set_enabled(bool)` for `EventCtx` and `UpdateCtx` which changes if the state of the current widget is "manually disabled".
- `is_enabled() -> bool` for all Context which returns false if this widget or any of its ancestors is "manually disabled".
- `LifeCycle::EnabledChanged(bool)` which is called every time when `is_enabled()` would return a different value than before.
- ~~the `EnabledIf` widget, which enables the enclosed widget if the provided closure returns `true`~~

~~It contains a breaking change: `LifeCycle::WidgetAdded` was changed to `LifeCycle::WidgetAdded{ initially_enabled: bool }`. This allows widgets to be disabled before the first call to `update()`~~.

~~I also added boxed versions of the methods that add children to a flex.~~

### CODE REVIEW

**Comment by cmyr** [comment]
Will do a big batch of reviews this weekend or early next week.

One initial piece of feedback: I'm not sure about the change to `WidgetAdded`. If a widget wants to be initially disabled it makes sense (to me) for it to disable itself while handling `WidgetAdded`; since this happens before the the widget has `layout` or `paint` called, I think it should be fine? If there's a particular problem you ran into that motivated this change I'm happy to hear about it. :)

**Comment by xarvic** [comment]
> One initial piece of feedback: I'm not sure about the change to WidgetAdded.
You are right, that was not necessary

File: druid/src/window.rs [hunk 1]
```diff
@@ -154,6 +154,18 @@ impl<T: Data> Window<T> {
...
+            let old = self.focus;
+            let new = self.widget_for_focus_request(focus_req);
+            // Only send RouteFocusChanged in case there's actual change
+            if old != new {
+                let event = LifeCycle::Internal(InternalLifeCycle::RouteFocusChanged { old, new });
+                self.lifecycle(queue, &event, data, env, false);
+                self.focus = new;
+            }
+        }
+
```
Comment (cmyr) [inline druid/src/window.rs:168]: Ugh, this is complicated. 😣

We will need to have handling in here to send the `EnabledChanged` event, and then we will need to update the focus chain; I think what makes sense then is to check if the currently focused widget still exists, and to only send a `FocusChanged` if it doesn't.

**Review by cmyr — CHANGES_REQUESTED** [review]
Thanks for taking this on! This is a very ambitious PR, and you're getting deep into the druid internals. Overall I think that the code is good, and I really appreciate the documentation.

I have two broad concerns:
- I think this PR is doing a bit too much, and I would really like to try and split out the *architectural changes* from both the display logic (widgets handling this new state) as well as from any addtional API that isn't directly related to the functionality at hand; this will let me review revisions of this PR much more quickly, which will hopefully let us get it landed more quickly. You can stash those other changes and we can review them separately, which should also be much simpler, since the logic for the widget stuff should (🤞) be more straightforward.
- I believe there are some quirks of the druid architecture that make specific aspects of this implementation slightly problematic. I think the big picture is good, but there are some tricky details that need to be sorted out; in particular I'm pretty sure we want to be dispatching events from the root of the tree, instead of injecting them in during the handling of the initial events.

In any case, I would encourage you to keep working on this; it's clear you've thought about it quite a bit and have a good grasp of the problem. If you have any questions please feel free to reach out here or on zulip, I'm happy to help however I can.

File: druid/src/widget/flex.rs [hunk 2]
```diff
@@ -419,6 +419,14 @@ impl<T: Data> Flex<T> {
...
     }
 
+    /// Builder-style variant of `add_boxed_child`.
+    ///
+    /// Convenient for assembling a group of widgets in a single expression.
+    pub fn with_boxed_child(mut self, child: Box<dyn Widget<T> + 'static>) -> Self {
+        self.add_boxed_flex_child(child, 0.0);
+        self
+    }
+
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/widget/flex.rs:429]: ok :)
If you think it could be useful i will open a separate PR.

File: druid/src/widget/radio.rs [hunk 3]
```diff
@@ -64,13 +64,15 @@ impl<T: Data + PartialEq> Widget<T> for Radio<T> {
     fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, _env: &Env) {
         match event {
             Event::MouseDown(_) => {
-                ctx.set_active(true);
-                ctx.request_paint();
+                if ctx.is_enabled() {
+                    ctx.set_active(true);
+                    ctx.request_paint();
+                }
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/widget/radio.rs:70]: Makes sense, i will remove it

File: druid/src/contexts.rs [hunk 4]
```diff
@@ -669,6 +708,22 @@ impl LifeCycleCtx<'_, '_> {
     pub fn register_for_focus(&mut self) {
         self.widget_state.focus_chain.push(self.widget_id());
     }
+
+    /// Sets the enabled state of a widget to false.
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/contexts.rs:712]: My idea was to create a api, which only allows useful things. WidgetAdded is indeed a funny special case, but it didn't seem like `set_enabled()` would make sense with any of the other LifeCycle events. 
It should also be possible to just delete changes to `set_enabled` happening during `EnabledChanged()`. I could make that change.

File: druid/src/contexts.rs [hunk 5]
```diff
@@ -271,6 +271,34 @@ impl_context_method!(
...
+        /// The widget can still be disabled if this value is `false`, use [`is_enabled`] to check if
+        /// the widget is currently enabled.
+        ///
+        /// If the value is `true` this widget will stay disabled at least until [`EventCtx::set_enabled`]
+        /// or [`UpdateCtx::set_enabled`] is called.
+        ///
+        /// [`is_enabled`]: #method.is_enabled
+        /// [`EventCtx::set_enabled`]: struct.EventCtx.html#method.set_enabled
+        /// [`UpdateCtx::set_enabled`]: struct.LifeCycleCtx.html#method.set_enabled
+        pub fn is_set_disabled(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/contexts.rs:286]: I thought it would be more complete, having the method. Maybe a user wants to access this information, but i also can't think of a good example how to make use of it. Your right, it probably just invites to user to use the `disabled` state in an unsound way. I will remove it.

File: druid/src/core.rs [hunk 6]
```diff
@@ -104,6 +104,12 @@ pub(crate) struct WidgetState {
     pub(crate) viewport_offset: Vec2,
 
     // TODO: consider using bitflags for the booleans.
+    pub(crate) set_enabled: bool,
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:107]: I think everything connected to enabled/disabled should either use "enabled" or "disabled" consistently. My initial thought was that `LifeCycle::DisabledChanged(false)` would sound strange, but your probably right, the user will search for disabled, since all widgets are enabled by default.

File: druid/src/core.rs [hunk 7]
```diff
@@ -104,6 +104,12 @@ pub(crate) struct WidgetState {
...
+
+    pub(crate) new_enabled: bool,
+
+    pub(crate) tree_enabled: bool,
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:111]: I'm not very good at naming things :)

File: druid/src/core.rs [hunk 8]
```diff
@@ -579,6 +585,43 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
         }
     }
 
+    fn update_enabled(
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:588]: Thanks for the review :)

Hmm, i haven't thought about that at all.
I see the point. It could get messy if we just send the event during another one.

I can think of two options to solve this:
- First: we add a fourth bool, something like `child-enabled-changed` and add a `InternalLifeCycle::RouteEnabledChanged()` event which gets send down the tree if `child-enabled-changed` is set or enabled was changed for the current widget. Then we could either update the focus chain directly or use children_changed.
- Second: we generalize `InternalLifeCycle::RouteWidgetAdded` to `InternalLifeCycle::RouteWidgetChanged` which would perform both tasks at once. This should work, since a widget can't receive both (`WidgetAdded` and `EnabledChanged`) at the same time. I like this version better since we save another `InternalLifeCycle` event, won't pollute the widgetstate with further dirty flags and also it feels like both events behave quite similar, since they are the only events which change the focus-chain. But i think for this to work, widgets should not be allowed to call `enabled_changed` from `Lifecycle::WidgetAdded`.

What are your thoughts about this?

File: druid/src/window.rs (hunk 1 above)
Comment (xarvic) [reply druid/src/window.rs:168]: Im not sure i understand what you mean. Are you talking about the situation when a widget calls `request_focus()` or the focus change issued in `update_enabled()`?

File: druid/src/core.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/core.rs:588]: The second option hadn't occurred to me, but I think it's *possible* it could work? I'd want to really think it through. One interesting thing about this approach is that we could potentially re-use the `children_changed` flag to indicate where we need to recurse in order to deliver the `FocusChanged` event.

Thinking this through more, though, I have my doubts. Basically: I think that this is exchanging superficial complexity ("one less event type! one less flag in `WidgetState`) for potentially deep combinational complexity, e.g. "now these two distinct actions interact in ways that are hard to reason about". For instance: imagine if, in response to the same event, one widget changes a child, and another widget calls `set_disabled`. Now we want to send `WidgetAdded`, but in the course of doing this we reset `children_changed`, and how we can't correctly route the `FocusChanged` event? I'm not sure if this is an actual concern; the point is more that I'm already getting a bit confused thinking about it.

So I think I prefer the first option. It's more verbose, but it is also simpler to reason about. Adding an additional `Internal` event really doesn't cost us anything meaningful.

A separate question is, in this world, what state we need to store in `WidgetState`. My current understanding is that we will need, at least (and without committing to these names, I'm just using what comes to me right now):

- `is_self_disabled`: whether *this specific widget* has been disabled
- `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`).
- `child_disabled_changed` a flag used when routing the actual event; when this is `false` we can stop recursing.
- `disabled_changed`: a flag which indicates that the *current* widget has been disabled. Actually: I think maybe this isn't a bool, but is more like an `Option<bool>`, to indicate the new state?

Oh no, this gets worse.

What happens when a widget and one of its descendants both change their disabled state in response to the same event, but they do it in different directions? For instance: what if a subtree was disabled, and the parent re-enables it, and at the same time one of the children explicitly disables itself?

Maybe this isn't that bad. With the actual `EnabledChanged` event, we recurse down, stopping at any child that has either `is_self_disabled` set, or that has `disabled_changed` as `Some(true)`, or whatever thing means "I'm becoming disabled"? In this latter case we would need to continue to descend only if the change was to also become disabled; otherwise we just set `is_self_disabled`, for that particular widget, but continue to notify children that they *are* now disabled.

So..... I think this is *okay* (in that it is not fundamentally broken) but it is definitely more complicated than I had initially thought, and I suspect there are edge cases I'm not thinking of yet. I would encourage you to try and think through these various permutations of possible states and orders, and I *definitely* think we are going to want some tests in place, if just to check that our understanding is correct.

File: druid/src/contexts.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/contexts.rs:712]: Yep that makes sense. `WidgetAdded` *does* make sense, though, so I think it's okay to just throw it in lifetime, too. Maybe someone wants to disable the current widget when the mouse leaves or something, who knows? 🤷

File: druid/src/core.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/core.rs:107]: I agree on both points, and also thought `DisabledChanged` sounds a bit weird but I don't mind it. We could also consider being verbose and adding `WasEnabled` and `WasDisabled` as separate events, although this would be a departure from current practice, although it might also be easier to reason about?

File: druid/src/window.rs (hunk 1 above)
Comment (cmyr) [reply druid/src/window.rs:168]: I'm also not sure I understand what I meant.

I think what I was concerned about, here, is the case where the widget that was currently focused becomes disabled, and thus is no longer in the focus chain. In this case we still need to send the `RouteFocusChanged` event, because we need to update every ancestor of that widget so that they update `has_focus`. Thinking about this more, we also need to send `FocusChanged` to the `WidgetPod` in question (to update its `is_focused` flag) but we *don't* (I think) need to send this message to the *actual widget*, which should have already been disabled, which implies a loss of focus. 🤔

File: druid/src/contexts.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/contexts.rs:286]: I think it might make sense here to be conservative; we don't include this method initially, and then add it if somebody can demonstrate a need.

File: druid/src/window.rs (hunk 1 above)
Comment (xarvic) [reply druid/src/window.rs:168]: My current Aproach is to set `focus_change_request` to `Resign` if a widget which contains the focused element disables it self. Doing this from the root would be difficult since we don't know which Widgets are allowed to be focused. Should I keep it that way?

File: druid/src/core.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:588]: I think the way you described should work, but i will definitely write some testcases this time :)

[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:588]: But i'm unsure about one thing:

> `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`)

My current Implementation has two fields: one for the own state (manually disabled) and one for the tree (ancestors, not including my self, disabled). I think we need both, since otherwise we lose the information if calling `set_enabled()` will change the "enabled state" of this widget.

Also `new_disabled` feels more robust than `enabled_changed`. The question is, is it possible to receive an event between calling `set_enabled()` and receiving `LifeCycle::EnabledChanged()` and if so should this event use the old or the new state?

File: druid/src/contexts.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/contexts.rs:286]: Ok

File: druid/src/core.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/core.rs:588]: > But i'm unsure about one thing:
> 
> > `is_disabled`: whether this widget is part of a disabled subtree (including the case where it is a subtree of one; `is_self_disabled` implies `is_disabled`)
> 
> My current Implementation has two fields: one for the own state (manually disabled) and one for the tree (ancestors, not including my self, disabled). I think we need both, since otherwise we lose the information if calling `set_enabled()` will change the "enabled state" of this widget.

I agree that we need two flags, one for the specific widget and one that includes parent state, and overrides the state of the specific widget; if a parent is disabled but we're not, we still *are*. I think we're on the same page about this.

> Also `new_disabled` feels more robust than `enabled_changed`. 

If you can make this work with `new_disabled` that's great. I guess maybe how I was thinking about this is as a two-part process; in the first part (when the widget calls `set_disabled()` we store information on *the pending change*, and then in the second part (when we traverse the tree to reconcile the changes and send the events) we clear the *pending* state and set the *actual* state. So in my thinking a widget isn't disabled when it calls `set_disabled`; it is disabled when the tree is traversed afterwards.

> The question is, is it possible to receive an event between calling `set_enabled()` and receiving `LifeCycle::EnabledChanged()` and if so should this event use the old or the new state?

The only event that it should be possible to receive here is `RouteWidgetAdded`; we need to always handle that before we do anything else. I think it should be fine to keep the new state in place while handling this event; immediately afterwards we will handle enabled/disabled, and then after that (if necessary) we will handle focus change.

File: druid/src/window.rs (hunk 1 above)
Comment (cmyr) [reply druid/src/window.rs:168]: I think your approach should work.

And it's a good point that we don't know what widgets can be focused; we only know what widgets participate in the focus chain, which is not *exactly* the same thing.

File: druid/src/core.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:107]: After thinking about this again `DisabledChanged()` is probably still the best option.

File: druid/src/contexts.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/contexts.rs:712]: Ok, then i delete any changes to disabled which happen during `EnabledChanged`.

File: druid/src/core.rs [hunk 9]
```diff
@@ -106,6 +106,25 @@ pub(crate) struct WidgetState {
     pub(crate) viewport_offset: Vec2,
 
     // TODO: consider using bitflags for the booleans.
+    // Any child of this widget changed the disabled state and should either receive
+    // LifeCycle::DisabledChanged or InternalLifeCycle::RouteDisabledChanged
```
Comment (cmyr) [inline druid/src/core.rs:110]:
```suggestion
    // `true` if a descendent of this widget changed its disabled state and should receive
    // LifeCycle::DisabledChanged or InternalLifeCycle::RouteDisabledChanged
```

File: druid/src/core.rs [hunk 10]
```diff
@@ -106,6 +106,25 @@ pub(crate) struct WidgetState {
...
+    pub(crate) children_disabled_changed: bool,
+
+    // Any of our ancestors are disabled. Therefore we are also disabled.
```
Comment (cmyr) [inline druid/src/core.rs:113]:
```suggestion
    // `true` if one of our ancestors is disabled (meaning we are also disabled).
```

File: druid/src/core.rs [hunk 11]
```diff
@@ -106,6 +106,25 @@ pub(crate) struct WidgetState {
...
+    pub(crate) ancestor_disabled: bool,
+
+    // explicitly disabled through the usage of EventCtx::set_disabled(true) on this widget.
```
Comment (cmyr) [inline druid/src/core.rs:116]:
```suggestion
    // `true` if this widget has been explicitly disabled.
    // A widget can be disabled without being *explicitly* disabled if an ancestor is disabled.
```

File: druid/src/core.rs [hunk 12]
```diff
@@ -106,6 +106,25 @@ pub(crate) struct WidgetState {
...
+    pub(crate) is_explicitly_disabled: bool,
+
+    // A buffer for the value of EventCtx::set_disabled() before this widget receives
+    // LifeCycle::DisabledChanged or InternalLifeCycle::RouteDisabledChanged
```
Comment (cmyr) [inline druid/src/core.rs:120]:
```suggestion
    // `true` if this widget has been explicitly disabled, but has not yet seen one of
    // LifeCycle::DisabledChanged or InternalLifeCycle::RouteDisabledChanged
```

(Is  this true?)

Also: it confuses me that this is a bool, since I also expect that we would want to know if a widget is newly *enabled*, e.g. it feels like there's a semantic difference between "nothing has changed" and "I have changed to become not-disabled"; both of these states are current represented by `false`.

I think I understand why this works, but I did find it confusing.

File: druid/src/core.rs [hunk 13]
```diff
@@ -106,6 +106,25 @@ pub(crate) struct WidgetState {
...
+    pub(crate) is_explicitly_disabled_new: bool,
+
+    // This widget get focused automatically.
+    // We need to store this here since the widget local focus-chain is cleared when the widget is
+    // disabled
```
Comment (cmyr) [inline druid/src/core.rs:125]:
```suggestion
    // `true` if this widget participates in the focus chain.
    // We store this as a bool because we need to know how to update the focus chain
    // when disabled/enabled state changes.
```

File: druid/src/core.rs [hunk 14]
```diff
@@ -1002,9 +1076,33 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
 
         // we need to (re)register children in case of one of the following events
+        if let LifeCycle::WidgetAdded = event {
+            // If this widget is not already present, autofocus is not set before the event,
+            // since we call register_for_focus in widget added.
+            // To be consistent with RouteWidgetAdded DisabledChanged and Route DisabledChange we
+            // insert our id at the first position.
+            if self.state.auto_focus {
+                self.state.focus_chain.insert(0, self.state.id);
+            }
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:1104]: Yes, that would also eliminate the need for `reset_focus_chain()`.

File: druid/src/tests/mod.rs [hunk 15]
```diff
@@ -350,6 +351,390 @@ fn focus_changed() {
...
+    let check_states = |name: &str, desired: [Option<bool>; 6]| {
+        if desired[0] != disabled_0.get()
+            || desired[1] != disabled_1.get()
+            || desired[2] != disabled_2.get()
+            || desired[3] != disabled_3.get()
+            || desired[4] != disabled_4.get()
+            || desired[5] != disabled_5.get()
+        {
+            panic!(format!(
+                "test \"{}\":\nexpected: {:?}\n got:      {:?}",
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/tests/mod.rs:576]: Ok :)

File: druid/src/window.rs [hunk 16]
```diff
@@ -182,6 +182,50 @@ impl<T: Data> Window<T> {
                 false,
             );
         }
+
+        // Update the disabled state if necessary
+        // Always do this before sending focus change, since this event updates the focus chain.
+        if self.root.state().tree_disabled_changed() {
+            let event = LifeCycle::Internal(InternalLifeCycle::RouteDisabledChanged);
+            self.lifecycle(queue, &event, data, env, false);
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/window.rs:190]: This could be what happened to me, when i was talking about the stackoverflow on Zulip.
Hmm :/ Calling `children_changed()` in `DisabledChanged` and `set_disabled()` in `RouteWidgetAdded`, will definitively result in an infinite loop. Maybe we should just warn about this for now and fix it in #1603?

File: druid/src/core.rs [hunk 17]
```diff
@@ -1002,9 +1076,33 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
+        }
+
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:1106]: I fixed some of them and added documentation to explain the rest.

File: druid/src/core.rs [hunk 18]
```diff
@@ -839,6 +858,9 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
         // Always merge even if not needed, because merging is idempotent and gives us simpler code.
         // Doing this conditionally only makes sense when there's a measurable performance boost.
         ctx.widget_state.merge_up(&mut self.state);
+
+        ctx.widget_state.children_disabled_changed |=
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:863]: My mistake :)

File: druid/src/core.rs [hunk 19]
```diff
@@ -908,7 +930,7 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
                     } else {
                         if self.state.children_changed {
                             self.state.children.clear();
-                            self.state.focus_chain.clear();
+                            self.state.reset_focus_chain();
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:934]: I changed the behavior again, because of your other reviews. Now the focus-chain collects the focus-chains from its children first and than prepends its own id if auto-focus is set. This works better, than appending the own value in the beginning since in case of `WidetAdded`, the auto_focus flag is is not set, then. Is this ok?

File: druid/src/window.rs [hunk 20]
```diff
@@ -182,6 +182,50 @@ impl<T: Data> Window<T> {
...
+        }
+
+        if let Some(focus_req) = widget_state.request_focus.take() {
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/window.rs:193]: I could do this as part of this PR if you like or should i just leave it there?

File: druid/src/core.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:934]: > Can you write down a quick summary of what the new expectations are around the focus chain? how does registration work, when does it happen, etc?

Should I write that in the docs?

File: druid/src/core.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:1106]: Hmmm :/ It wasn't as easy as i thought. I think some of this complexity is necessary. 

- The ID of this widget should be the first in our focus-chain
- we should add our ID only if auto_focus is set
- auto_focus is not set until after `WidgetAdded`
- we only need to add the value if children_changed or children_disabled_changed is set, otherwise we duplicate the values.
- some_one could call `ctx.children_changed()` while routing `WidgetAdded`, which makes no sense, but therefore after the call we don't know if the value was set

The current implementation always adds the own id when the focus_chain is cleared and focus_chain is set. Only in `WidgetAdded` the id is always added after the event.
I don't know if there is a better way to satisfy the the conditions i mentioned?

File: druid/src/core.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/core.rs:931]: It's a good question where this should be. Maybe in the docs for like `RouteWidgetAdded`? Or otherwise just as a comment in the `WidgetState` struct, where all the fields are defined?

File: druid/src/core.rs [hunk 21]
```diff
@@ -957,6 +977,25 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
+                        self.state.reset_focus_chain();
+                        extra_event = Some(LifeCycle::DisabledChanged(self.state.is_disabled()));
+                        //Each widget needs only one of DisabledChanged and RouteDisabledChanged
+                        false
+                    } else if self.state.children_disabled_changed {
+                        self.state.reset_focus_chain();
+                        true
+                    } else {
+                        false
+                    }
```
Comment (cmyr) [inline druid/src/core.rs:997]: okay trying to wrap my head around this.

Can we simplify this?

- *always* clear the focus chain
- *always* fixup the focus chain afterwards?

I'm also curious about the logic here. I would expect it to be true that in order to see this event, *either* we have been explicitly disabled, *or* some of our children are, and the final branch here should be unreachable?

**Review by cmyr — COMMENTED** [review]
A couple additional notes, I'm wondering if there are ways we can simplify the control flow here? Where possible, I really want to design mechanisms that don't rely on a bunch of special cases. In particular I'm wondering if we might be able to have fewer branches and conditionals, and just *always* reset the focus chain before dispatching these events, and then *always* fixing the focus chain up, afterwards.

File: druid/src/core.rs (hunk 21 above)
Comment (cmyr) [reply druid/src/core.rs:997]: Basically: looking back through this method, I am struck by how simple it is to handle the focus_chain stuff currently, and I feel like it should be possible to have the new logic be similarly simple? This all comes down to understanding our invariants; what guarantees can we make about our state before we recurse, and what guarantees can we expect afterwards?

Comment (xarvic) [reply druid/src/core.rs:997]: If we always clear the focus_chain we need to route this event to all widgets in the window. I could change it if you think that's better. 
In my implementation i always clear the focus-chain if i recurse the event or send the extra event.

File: druid/src/core.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:931]: Then i will write it in WidgetState

File: druid/src/core.rs [hunk 22]
```diff
@@ -957,6 +977,25 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
                     f.call(&self.state);
                     true
                 }
+                InternalLifeCycle::RouteDisabledChanged => {
```
[the remark this answers is not in the document]
Comment (xarvic) [reply druid/src/core.rs:980]: Ok :)

File: druid/src/core.rs [hunk 23]
```diff
@@ -980,6 +1019,27 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
+
+                if was_disabled != self.state.is_disabled() {
+                    // In case we change but none of our children does we still need to update the
+                    // focus-chain
+                    self.state.reset_focus_chain();
+                    extra_event = Some(LifeCycle::DisabledChanged(self.state.is_disabled()));
+                } else if self.state.children_disabled_changed {
+                    self.state.reset_focus_chain();
+                    extra_event =
+                        Some(LifeCycle::Internal(InternalLifeCycle::RouteDisabledChanged));
```
Comment (cmyr) [inline druid/src/core.rs:1036]: okay I think I also missed this in my first pass, and I'm having trouble wrapping my head around it.

When do we want to switch from `DisabledChanged` back to `RouteDisabledChanged`?

File: druid/src/window.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply druid/src/window.rs:193]: It's fine here I think.

File: druid/src/core.rs (hunk 23 above)
Comment (xarvic) [reply druid/src/core.rs:1036]: I probably should write that in the code

Comment (cmyr) [reply druid/src/core.rs:1036]: Do we actually need to route in this case? The child was disabled previously and remains disabled. Or is this about ensuring that the internal state is correctly reset?

Comment (xarvic) [reply druid/src/core.rs:1036]: Oh, i forgot about that. It should be possible to leave the state dirty since it is already disabled. Even if is_explicitly_disabled is false and is_explicitly_disabled_new is true, this should get handled correctly and the widget would stay disabled when it receives `DisabledChanged(false)`. But it feels like moving the complexity to another place, because now it is possible to receive events while the state is dirty. And i can't tell how this would affect other things.

File: druid/src/core.rs [hunk 24]
```diff
@@ -1130,10 +1216,21 @@ impl WidgetState {
             cursor_change: CursorChange::Default,
             cursor: None,
             sub_window_hosts: Vec::new(),
+            is_explicitly_disabled_new: false,
             text_registrations: Vec::new(),
+            update_focus_chain: true,
```
Comment (cmyr) [inline druid/src/core.rs:1221]: do we need to track this? Can we just always update the focus change after any `ChildrenAdded` or `DisabledChanged` event?

File: druid/src/core.rs [hunk 25]
```diff
@@ -908,11 +925,29 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
                 }
+                InternalLifeCycle::RouteDisabledChanged => {
+                    let was_disabled = self.state.is_disabled();
+
+                    self.state.is_explicitly_disabled = self.state.is_explicitly_disabled_new;
+
+                    if was_disabled != self.state.is_disabled() {
+                        // In case we change but none of our children we still need to update the
+                        // focus-chain
+                        self.state.update_focus_chain = true;
```
Comment (cmyr) [inline druid/src/core.rs:940]: I would be happy to have less update logic here and just *always* rebuild the focus chain after this event; it's harder for us to screw that up later when we make some other little change.

File: druid/src/core.rs [hunk 26]
```diff
@@ -980,13 +1017,37 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
+
+                // the change direction (true -> false or false -> true) of our parent and ourself
+                // is always the same, or we dont change at all, because we stay disabled if either
+                // we or our parent are disabled.
+                if was_disabled != self.state.is_disabled() {
+                    self.state.update_focus_chain = true;
+                    true
+                } else {
+                    false
+                }
```
Comment (cmyr) [inline druid/src/core.rs:1034]: yea if we didn't need to update this manually we could just write,

```rust
was_disabled != self.state.is_disabled()
```

File: druid/src/event.rs [hunk 27]
```diff
@@ -265,6 +265,13 @@ pub enum LifeCycle {
     /// [`Size`]: struct.Size.html
     /// [`Widget::layout`]: trait.Widget.html#tymethod.layout
     Size(Size),
+    /// Called when the Disabled state of the widgets is changed.
+    ///
+    /// [`is_disabled`](struct.EventCtx.html#method.is_disabled) returns if the widget is disabled.
+    ///
+    /// [`set_disabled`](struct.EventCtx.html#method.set_disabled) to change the widgets disabled
+    /// state.
```
Comment (cmyr) [inline druid/src/event.rs:273]: Project style is to avoid inline links, it makes it harder to read the docs when you're in the actual code.
```suggestion
    /// To check if a widget is disabled, see [`is_disabled`].
    ///
    /// To change a widget's disabled state, see [`set_disabled`].
    ///
    /// [`is_disabled`]: crate::EventCtx::is_disabled
    /// [`set_disabled`]: crate::EventCtx::set_disabled
```

File: druid/src/event.rs [hunk 28]
```diff
@@ -274,6 +281,16 @@ pub enum LifeCycle {
...
     HotChanged(bool),
+    /// This is called when the widget-tree changes and druid wants to rebuild the
+    /// Focus-chain.
+    ///
+    /// It is the only place from witch [`register_for_focus`] should be called.
+    /// By doing so the widget can get focused by other widgets using [`focus_next`] or [`focus_prev`].
+    ///
+    /// [`register_for_focus`](struct.LifecycleCtx.html#method.register_for_focus)
+    /// [`focus_next`](struct.LifecycleCtx.html#method.register_for_focus)
+    /// [`focus_prev`](struct.LifecycleCtx.html#method.register_for_focus)
```
Comment (cmyr) [inline druid/src/event.rs:292]: Let's use the new intra-doc links style:
```suggestion
    /// [`register_for_focus`]: crate::LifecycleCtx::register_for_focus
    /// [`focus_next`]: crate::EventCtx::focus_next
    /// [`focus_prev`]: crate::EventCtx::focus_prev
```

**Review by cmyr — COMMENTED** [review]
Okay, this looks like it's very close. My one remaining thought is whether we can remove some of the logic around deciding when to update the focus chain, and just always do that after any `WidgetAdded` or `DisabledChanged` event?

File: druid/src/core.rs (hunk 24 above)
Comment (xarvic) [reply druid/src/core.rs:1221]: Ok :)

File: druid/src/core.rs (hunk 25 above)
Comment (xarvic) [reply druid/src/core.rs:940]: Ok

**Comment by SecondFlight** [comment]
While you're updating this PR, could you add this feature to the [changelog](https://github.com/linebender/druid/blob/master/CHANGELOG.md)?

**Comment by xarvic** [comment]
I will add it to the changelog with the next batch of fixes. I hope that doesn't take too long

**Review by cmyr — APPROVED** [review]
Okay I still would not be surprised if we run into some small issues with this going forward, but I think it's worth getting it merged now and then we will see what shakes out. Thank you @xarvic for taking this on, it was a complicated feature and I'm happy with how it's looking!
